x-pack/plugins/siem/server/lib/detection_engine/signals/types.ts

```diff
@@ -34,7 +34,10 @@ export type SignalsStatusRestParams = Omit<SignalsStatusParams, 'signalIds'> & {
 };
 
 export type SignalsQueryRestParams = SignalQueryParams;
-
+export interface EventSource {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  [field: string]: any;
+}
```

Comment:
This is turning everything into an `any` all over the place once you extend it.

I don't think this is what you really want to be doing. It's going to cause so many type mismatches and basically render your typings with:

```ts
export interface SignalSource extends EventSource {
```

to be an `any` and have no shape. That and it will just mess you up later when you want to give stronger shapes or a more exhaustive strong type to something that extends it. Lastly, `EventSource` is an overloaded type from ES6:
https://developer.mozilla.org/en-US/docs/Web/API/EventSource

So be careful with naming things that ;-)

I would suggest instead to remove this and use this below as `SignalSource`:

```ts
export interface SignalSource {
  '@timestamp': string;
  event?: {};
  [key: string]: unknown;
  signal?: {
    parent: Ancestor;
    ancestors: Ancestor[];
  };
}
```

That keeps your strong types such as optional `event` and allows us to fill in specific types but if it does not find one then it will fall back on `unknown` and we can manage that type separately.

In other areas of code you just have to use an `as cast` like so here:
x-pack/plugins/siem/server/lib/detection_engine/signals/filter_events_with_list.ts

```ts
      const valuesOfGivenType = eventSearchResult.hits.hits
        .map(searchResultItem => {
          return searchResultItem._source[exceptionItem.field];
        })
        .filter(searchResultItem => searchResultItem != null);
      const valueSet = new Set<string>(valuesOfGivenType as string[]); // make them small
```

```ts
? !badSet.has(item._source[exceptionItem.field] as string)
```

Notice I now remove the ugly lodash `get` call as well? Nicer code, less lodash, and less any. I will repeat these commented parts below with other tips on io-ts things people might not know about.

Reply:
Originally the `SignalSource` type had `[key: string]: SearchTypes;` but I think if I change that to what you mention `[key: string]: unknown;` I should be fine. I remember changing the type to `any` because of the nesting structure of the search results (which is actually why I had to use lodash get too). e.g.

```node.js
{
  _source: {
    host: {
      ip: '0.0.0.0'
    }
  }
}
```

I can't access `item._source[exceptionItem.field]` if `exceptionItem.field='host.ip'`, which is why I was using the lodash `get` because it is able to expand flattened key structure.

I'll remove the `EventSource` type and just update the `[key: string]` to `unknown` instead of `SearchTypes`. Not sure if it's possible to remove the lodash get though.

Reply:
Fixed (I think) [here](10a7f3901bdcded6a17d916a996c37b4b3b6aeb9) and [here](278451498f112a8f16c6109b4f997f8ae1c6a7ca)